cufinufftpy/cufinufft.py

```diff
@@ -0,0 +1,198 @@
+
+        # Initialize the plan for this instance
+        self._plan()
+
+    def default_opts(self, nufft_type, dim):
+        """
+        Generates a cufinufft opt struct of the dtype coresponding to plan.
+
+        :param finufft_type: integer 1, 2, or 3.
+        :param dim: Integer dimension.
+
+        :return: nufft_opts structure.
+        """
+
+        nufft_opts = self.Nufft_Opts()
+
+        ier = self._default_opts(nufft_type, dim, nufft_opts)
+
+        if ier != 0:
+            raise RuntimeError('Configuration not yet implemented.')
+
+        return nufft_opts
+
+    def _plan(self):
+        """
+        Internal method to initialize plan struct and call low level make_plan.
+        """
+
+        # Initialize struct
+        plan = self.CufinufftPlan()
+        plan.opts = self.opts
+
+        ier = self._make_plan(self._finufft_type,
+                              self.dim,
+                              self.modes,
+                              self.isign,
+                              self.ntransforms,
+                              self.tol,
+                              1,
+                              plan)
+
+        if ier != 0:
+            raise RuntimeError('Error creating plan.')
+
+        self.plan = plan
+
+    def set_nu_pts(self, M, kx, ky=None, kz=None):
+        """
+        Sets non uniform points of the correct dtype.
+
+        Note kx, ky, kz are required for 1, 2, and 3
+        dimensional cases respectively.
+
+        :param M: Number of points
+        :param kx: Array of x points.
+        :param ky: Array of y points.
+        :param kz: Array of z points.
+        """
+
+        if not (kx.dtype == ky.dtype == kz.dtype == self.dtype):
+            raise TypeError("cifinufft plan.dtype and "
+                            "kx, ky, kz dtypes do not match.")
+
+        kx = kx.ptr
+
+        if ky is not None:
+            ky = ky.ptr
+
+        if kz is not None:
+            kz = kz.ptr
+
+        ier = self._set_nu_pts(M, kx, ky, kz, 0, None, None, None, self.plan)
+
+        if ier != 0:
+            raise RuntimeError('Error setting non-uniform points.')
+
+    def execute(self, c, fk):
+        """
+        Executes plan. Note the IO orientation of `c` and `fk` are
+        determined by plan type.
+
+        In type 1, `c` is input, `fk` is output, while
+        in type 2, 'fk' in input, `c` is output.
+
+        :param c: Real space array in 1, 2, or 3 dimensions.
+        :param fk: Fourier space array in 1, 2, or 3 dimensions.
+        """
+
+        if not c.dtype == fk.dtype == self.complex_dtype:
+            raise TypeError("cufinufft execute expects {} dtype arguments "
+                            "for this plan. Check plan and arguments.".format(
+                                self.complex_dtype))
+
+        ier = self._exec_plan(c.ptr, fk.ptr, self.plan)
+
+        if ier != 0:
+            raise RuntimeError('Error executing plan.')
+
+    def destroy(self):
```

Comment:
Should this just be `__del__`?

Reply:
Good point, I think I used `__del__` when I was hacking around in ASPIRE. I forgot :). Changed to `__del__`. I removed `plan.destroy()` from the pytests and confirmed they are still running.

In practice we may find that we do not want to wait for python to clean these up , but that is an optimization. That is, I'm not sure how long after the plan instance ref count is zero these will hang around before `__del__` is called.  If it is immediate that's great, otherwise, you can imagine having a loop...  I think it will be pretty obvious if a problem..

Reply:
On the other hand, the user can just call `del plan` in that case, right? This just allows for an automatic (albeit potentially problematic) fallback.